### PR TITLE
remove FileCodebase and some weeds

### PR DESCRIPTION
**Pull request by aryairani**
In addition to removing v1 `FileCodebase` support, this PR decouples many module dependencies that I felt were making refactors difficult.

Apart from a couple of tweaked output messages, there shouldn't be any functionality changes here.

This will merge-conflict pretty hard with #2365 unfortunately, but I'm happy to help with reconciling them.

**Things the PR did**
- pulled these into their own modules:
    - Codebase (types) into `Codebase.Type`; `Codebase` module re-exports them
    - `CodebasePath`
    - `Codebase.Branch.Merge`, 
    - `Codebase.Branch.Names` (Branch shouldn't depend on `Names`)
    - `Codebase.BuiltinAnnotation`
    - `Codebase.Causal.FoldHistory`
    - `Codebase.CodeLookup.Util` (remove `CodeLookup` dependency on `UnisonFile`)
    - `Codebase.DataDeclaration.ConstructorId` trying to use this alias in relevant places
    - `Codebase.Init`
    - `CodebasePath`, the `FilePath` alias.
    - `Lexer.Pos` (because `AnnotatedText`, `Range`, `Parser.Ann` shouldn't depend on the whole lexer)
    - `Names.ResolutionResult` (`Type` shouldn't depend on `Names`)
    - `PrettyPrintEnv.FQN` (contains `elideFQN` logic)
    - `PrettyPrintEnv.Names` (pulled out references to `Names` from `PPE`)
    - `PPE.Util` not 100% sure what's happening here
    - `PrettyPrintEnvDecl` pull this data type and supporting functions into separate module
    - `Path.Parse` (`Path` shouldn't depend on the lexer)
    - `Path.Convert`, `Path.Parse` into `Unison.Util.Convert`
    - `Parser.Ann` (`Codebase`, etc. etc. shouldn't depend on `Parser`)
    - `Reference.Util` (`Reference` shouldn't depend on `ABT`)
    - `Referent'` (`SyntaxText`/`ColorText` shouldn't depend on `Reference`)
        - `Referent` module re-exports stuff from `Referent'`, hard-coded to `Reference`.
    - removed `SyntaxText.SyntaxText`
    - `TermEdit.Typing` (because `TermEdit` shouldn't depend on the full typechecker)
    - `UnisonFile` / `TypecheckedUnisonFile` (types into `UnisonFile.Type`)
    - `UnisonFile.Env`
    - `UnisonFile.Error`
    - `UnisonFile.Names`
    - `Var.WatchKind`
    - `Var.refNamed`
- Deleted unused `Codebase.Classes` typeclasses wip
- Deleted unused `Unison.Util.Menu`, ancient modal stuff
- deletes the `Runtime.needsContainment` field and related logic.
- changed `bindNames` to `bindReferences` in some cases; `bindNames` remains in `.Names` compatibility module
- move `Unison.Var.refNamed` into `Unison.Term`
- tweaked GitError to separate obviously codebase-format specific errors from non-obviously-corbase-format-specific errors (see `Codebase.Type.GitError`)
- tweaked `Reflog.Entry` to support anything that's coercible to `Unison.Hash`, but also changed its kind
- removed `DebugBranchHistoryI` input because I didn't want to maintain its implementation
- removed `ShortBranchHash` dependency on `Hash`, and let it work on anything coercible to `Hash`.  P.S./todo This class should be called ShortHash; the existing ShortHash is really a ShortReference! (because it contains extras to do with cycles, etc.)
- removed SyntaxText dependency on Reference
- cleaned up `Path` somewhat

**Things I debated**
- in spots (like `Codebase`, `Codebase.Init`, `UnisonFile`) where I moved the actual type out of the file, I wanted to separate the definition from the various util functions.  I'd be happy to move the util functions instead.
- should I restore `data CodebaseFormat = V2` in Main/GetArgs? (like in Test.Ucm) or leave it deleted until we have multiple formats?

**Things I held off on / maybe later**
- move `DD.updateDependencies`
- splitting up `Path` into the billion different components
    - e.g. `Path` becomes `RelativePath`, `Path'` becomes `Path`, and `Absolute` wraps `RelativePath`
    - `Split` / `Split'` maybe rename to `Path.NonEmpty`?
- rename `Branch.getPatch` / `.getMaybePatch`
- split out `Name.Parse`, `Name.Convert` type classes, substitute a lot of specific functions like `Path.hqSplitFromName'` with `Convert.parse`.
- rename `ShortHash` to `ShortReference`, and `ShortBranchHash` to `ShortHash`

### CODE REVIEW

File: unison-core/src/Unison/Referent.hs [hunk 1]
```diff
@@ -135,7 +123,3 @@ fold :: (r -> a) -> (r -> Int -> ConstructorType -> a) -> Referent' r -> a
 fold fr fc = \case
   Ref' r -> fr r
   Con' r i ct -> fc r i ct
-
-instance Hashable Referent where
-  tokens (Ref r) = [H.Tag 0] ++ H.tokens r
```
Comment (pchiusano) [inline unison-core/src/Unison/Referent.hs:140]: Hmm, where'd this `Hashable` instance go? Just want to make sure we aren't accidentally introducing any hashing changes.

Comment (pchiusano) [reply unison-core/src/Unison/Referent.hs:140]: Oh, nevermind, I see, `Referent'` was just split out into its own module.

File: unison-core/src/Unison/Type.hs [hunk 2]
```diff
@@ -61,18 +61,17 @@ bindExternal
...
-bindNames keepFree ns0 t = let
-  ns = Names.Names ns0 mempty
+bindReferences keepFree ns t = let
   fvs = ABT.freeVarOccurrences keepFree t
-  rs = [(v, a, Names.lookupHQType (Name.convert $ Name.fromVar v) ns) | (v,a) <- fvs ]
-  ok (v, a, rs) = if Set.size rs == 1 then pure (v, Set.findMin rs)
-                  else Left (pure (Names.TypeResolutionFailure v a rs))
+  rs = [(v, a, Map.lookup (Name.fromVar v) ns) | (v, a) <- fvs]
+  ok (v, _a, Just r) = pure (v, r)
+  ok (v, a, Nothing) = Left (pure (Names.TypeResolutionFailure v a mempty))
```
Comment (pchiusano) [inline unison-core/src/Unison/Type.hs:74]: Where's `Type.bindNames` live now?

File: unison-core/src/Unison/Type/Names.hs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  -> Names.Names0
+  -> Type v a
+  -> Names.ResolutionResult v a (Type v a)
+bindNames keepFree ns0 t = let
+  ns = Names.Names ns0 mempty
+  fvs = ABT.freeVarOccurrences keepFree t
+  rs = [(v, a, Names.lookupHQType (Name.convert $ Name.fromVar v) ns) | (v,a) <- fvs ]
+  ok (v, a, rs) = if Set.size rs == 1 then pure (v, Set.findMin rs)
+                  else Left (pure (Names.TypeResolutionFailure v a rs))
+  in List.validate ok rs <&> \es -> bindExternal es t
```
Comment (pchiusano) [inline unison-core/src/Unison/Type/Names.hs:33]: Oh, here we go. Disregard earlier question.

**Review by pchiusano — APPROVED** [review]
Cool. Left a couple minor questions.

Can you work with @hojberg and @rlmark on refreshing https://github.com/unisonweb/unison/pull/2365 which I'm guessing will have some conflicts with `Main`

I think it would be good to document conventions for how modules get broken up, and then start doing that consistently, particularly for new code or when you're touching existing code.

Like if the convention is: `Foo` has the type and instances, and `Foo.Util` has utility functions (and re-exports `Foo`?) then do that consistently. If the convention is that `Foo.Type` has the type and instances, and `Foo` has the utility functions (and re-exports `Foo.Type`?) then do that consistently.

Having a "1 type per module, unless it's a util module that's about re-exporting" guideline might be a good thing to start doing when writing code in the first place.

Possibly relevant: https://www.haskellforall.com/2021/05/module-organization-guidelines-for.html

File: unison-core/src/Unison/Type.hs (hunk 2 above)
Comment (pchiusano) [reply unison-core/src/Unison/Type.hs:74]: Nvm, I see.

**Comment by aryairani** [comment]
> I think it would be good to document conventions

Yeah, I’ll write something up, to answer these questions, and maybe tweak what’s here to match.

> Having a "1 type per module, unless it's a util module that's about re-exporting" guideline might be a good thing to start doing when writing code in the first place.

Yeah.

> Possibly relevant: https://www.haskellforall.com/2021/05/module-organization-guidelines-for.html

Cool, I’ll check it out.

**Pull request merged**